include/boost/url/impl/decode.hpp

```diff
@@ -0,0 +1,160 @@
+//
+// Copyright (c) 2019 Vinnie Falco (vinnie.falco@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_IMPL_DECODE_HPP
+#define BOOST_URL_IMPL_DECODE_HPP
+
+#include <boost/url/detail/except.hpp>
+#include <boost/url/detail/except.hpp>
+#include <boost/url/decode_view.hpp>
+#include <boost/url/grammar/hexdig_chars.hpp>
+#include <boost/url/grammar/type_traits.hpp>
+#include <boost/url/grammar/charset.hpp>
+#include <boost/assert.hpp>
+#include <boost/static_assert.hpp>
+
+namespace boost {
+namespace urls {
+
+namespace detail {
+template<class CharSet>
+result<std::size_t>
+validate_encoding(
+    string_view s,
+    CharSet const& allowed,
+    decode_opts const& opt) noexcept
+{
+    // CharSet must satisfy is_charset
+    BOOST_STATIC_ASSERT(
+        grammar::is_charset<CharSet>::value);
+
+    // can't have % in charset
+    BOOST_ASSERT(! allowed('%'));
+
+    // we can't accept plus to space if '+' is not allowed
+    BOOST_ASSERT(! opt.plus_to_space || allowed('+'));
+
+    std::size_t n = 0;
+    char const* it = s.data();
+    char const* end = it + s.size();
+    while(it != end)
+    {
+        if( ! opt.allow_null &&
```

Comment:
shouldn't the decision to allow null be in the CharSet? Otherwise we can't use `find_if_not`...

Reply:
It can also be in the charset. The difference is `allow_null` will check if this is null before and after decoding, while removing it from the charset only forbids it before decoding. So we would still need the option `allow_null` to be available.

We don't use  `find_if_not` but we could use it to find the next '%'/'\0' and memcopy the rest. I don't know how much we would gain by doing that, but I think it's worth investing more time on the API for now.

Reply:
Right, I'm not saying to use `find_if_not` now but I am pointing out that if we want to use it in the future we need to be careful with the API